src/machine_learning/loss_function/negative_log_likelihood.rs

```diff
@@ -0,0 +1,77 @@
+// Negative Log Likelihood Loss Function
+//
+// The `neg_log_likelihood` function calculates the Negative Log Likelyhood loss,
+// which is a loss function used for classification problems in machine learning.
+//
+// ## Formula
+//
+// For a pair of actual and predicted values, represented as vectors `y_true` and
+// `y_pred`, the Negative Log Likelihood loss is calculated as:
+//
+// - loss = `-y_true * log(y_pred) - (1 - y_true) * log(1 - y_pred)`.
+//
+// It returns the average loss by dividing the `total_loss` by total no. of
+// elements.
+
+pub fn neg_log_likelihood(y_true: &[f64], y_pred: &[f64]) -> Option<f64> {
+    // Checks if the length of the actual and predicted values are equal
+    if y_true.len() != y_pred.len() || y_pred.is_empty() || y_true.is_empty() {
+        return None;
+    }
+    // Checks values are between 0 and 1
+    for (p, a) in y_pred.iter().zip(y_true.iter()) {
+        if *p < 0.0 || *p > 1.0 || *a < 0.0 || *a > 1.0 {
+            return None;
+        }
+    }
+
+    let mut total_loss: f64 = 0.0;
+    for (p, a) in y_pred.iter().zip(y_true.iter()) {
+        let loss: f64 = -a * p.ln() - (1.0 - a) * (1.0 - p).ln();
+        total_loss += loss;
+    }
+    Some(total_loss / (y_pred.len() as f64))
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn test_neg_log_likelihood() {
+        let actual_values: Vec<f64> = vec![1.0, 0.0, 1.0];
+        let predicted_values: Vec<f64> = vec![0.9, 0.1, 0.8];
+        assert_eq!(
+            neg_log_likelihood(&actual_values, &predicted_values),
+            Some(0.14462152754328741)
+        );
+    }
+
+    #[test]
+    fn test_neg_log_likelihood_invalid_length0() {
+        let actual_values: Vec<f64> = vec![1.0, 0.0, 1.0];
+        let predicted_values: Vec<f64> = vec![0.9, 0.1];
+        assert_eq!(neg_log_likelihood(&actual_values, &predicted_values), None);
+    }
+
+    #[test]
+    fn test_neg_log_likelihood_invalid_length1() {
+        let actual_values: Vec<f64> = vec![1.0, 0.0];
+        let predicted_values: Vec<f64> = vec![0.9, 0.1, 0.8];
+        assert_eq!(neg_log_likelihood(&actual_values, &predicted_values), None);
+    }
+
+    #[test]
+    fn test_neg_log_likelihood_invalid_values() {
+        let actual_values: Vec<f64> = vec![1.0, 0.0, 1.0];
+        let predicted_values: Vec<f64> = vec![1.1, 0.1, 0.8];
+        assert_eq!(neg_log_likelihood(&actual_values, &predicted_values), None);
+    }
+
+    #[test]
+    fn test_neg_log_likelihood_empty_prediction() {
+        let actual_values: Vec<f64> = vec![1.0, 0.0, 1.0];
+        let predicted_values: Vec<f64> = vec![];
+        assert_eq!(neg_log_likelihood(&actual_values, &predicted_values), None);
+    }
```

Comment:
```suggestion
    macro_rules! test_with_wrong_inputs {
        ($($name:ident: $inputs:expr,)*) => {
            $(
                #[test]
                fn $name() {
                    let (values_a, values_b) = $inputs;
                    assert!(neg_log_likelihood(&values_a, &values_b).is_none());
                    assert!(neg_log_likelihood(&values_b, &values_a).is_none());
                }
            )*
        }
    }

    test_with_wrong_inputs! {
        different_length: (vec![0.9, 0.0, 0.8], vec![0.9, 0.1]),
        different_length_one_empty: (vec![], vec![0.9, 0.1]),
        value_greater_than_1: (vec![1.1, 0.0, 0.8], vec![0.1, 0.2, 0.3]),
        value_greater_smaller_than_0: (vec![0.9, 0.0, -0.1], vec![0.1, 0.2, 0.3]),
        empty_input: (vec![], vec![]),
    }
```

Reply:
@jkauerl please have a look here.

Reply:
I tried implementing the macro and when I did the testing, it took longer than the allowed time.

Reply:
Could you elaborate more on that? This sounds really strange. There should be no impact on the performance.

Reply:
Oh, if there should not be an impact on performance then it might be the computer I am using.